src/coreComponents/mesh/utilities/CIcomputationKernel.hpp

```diff
@@ -0,0 +1,191 @@
+ * ------------------------------------------------------------------------------------------------------------
+ */
+
+
+/**
+ * @file CIcomputationKernel.hpp
+ */
+#ifndef GEOSX_MESH_UTILITIES_CICOMPUTATIONKERNEL_HPP_
+#define GEOSX_MESH_UTILITIES_CICOMPUTATIONKERNEL_HPP_
+
+#include "common/DataTypes.hpp"
+#include "common/TimingMacros.hpp"
+#include "finiteElement/FiniteElementDispatch.hpp"
+#include "common/GEOS_RAJA_Interface.hpp"
+#include "mesh/EmbeddedSurfaceSubRegion.hpp"
+#include "mesh/CellElementSubRegion.hpp"
+#include "mesh/ElementType.hpp"
+
+
+namespace geosx
+{
+
+template< typename FE_TYPE >
+class CIcomputationKernel
+{
+public:
+  CIcomputationKernel( FE_TYPE const & finiteElementSpace,
+                       NodeManager const & nodeManager,
+                       CellElementSubRegion const & elementSubRegion,
+                       EmbeddedSurfaceSubRegion & embeddedSurfSubRegion ):
+    m_finiteElementSpace( finiteElementSpace ),
+    m_elementType( elementSubRegion.getElementType() ),
+    m_X( nodeManager.referencePosition() ),
+    m_elemsToNodes( elementSubRegion.nodeList().toViewConst() ),
+    m_fracturedElems( elementSubRegion.fracturedElementsList().toViewConst()),
+    m_cellsToEmbeddedSurfaces( elementSubRegion.embeddedSurfacesList().toViewConst() ),
+    m_normalVector( embeddedSurfSubRegion.getNormalVector().toViewConst()),
+    m_fracCenter( embeddedSurfSubRegion.getElementCenter().toViewConst()),
+    m_fractureSurfaceArea( embeddedSurfSubRegion.getElementArea().toViewConst() ),
+    m_connectivityIndex( embeddedSurfSubRegion.getConnectivityIndex() )
+  {}
+
+  static constexpr int numNodesPerElem = FE_TYPE::maxSupportPoints;
+
+  static constexpr int numSamplingPoints = FE_TYPE::numSamplingPoints;
+
+  struct StackVariables
+  {
+public:
+
+    /// Constructor.
+    GEOSX_HOST_DEVICE
+    StackVariables():
+      xLocal(),
+      samplingPointCoord()
+    {}
+
+    /// C-array stack storage for element local the nodal positions.
+    real64 xLocal[ numNodesPerElem ][ 3 ];
+
+    real64 samplingPointCoord[3];
+  };
+
+
+  template< typename POLICY,
+            typename KERNEL_TYPE >
+  static
+  void launchCICompuationKernel( KERNEL_TYPE & kernelComponent )
+  {
+    GEOSX_MARK_FUNCTION;
+    forAll< POLICY >( kernelComponent.m_fracturedElems.size(),
+                      [=] GEOSX_HOST_DEVICE ( localIndex const i )
+    {
+
+      localIndex k = kernelComponent.m_fracturedElems[i];
+
+      StackVariables stack;
+      kernelComponent.setup( k, stack );
+
+      real64 averageDistance = 0.0;
+      for( integer np=0; np<numSamplingPoints; ++np )
+      {
+        kernelComponent.samplingPointCoord( np, stack );
+        averageDistance += kernelComponent.computeDistance( k, stack.samplingPointCoord );
+      }
+      averageDistance /= numSamplingPoints;
+      kernelComponent.setConnectivityIndex( k, averageDistance );
+    } );
```

Comment:
If this is the part that you wanted to thread, I don't think you would gain much performance by using multiple threads for this, you may actually loose performance by requiring using atomics and by having idle threads.